### ISSUE TITLE
Q: Improving scheduler and make it useful for all apps

### ISSUE THREAD

**Issue by shankardevy** [evercam/evercam-server#86]
_Goal:_ Make a generic scheduler for use by any app that requires polling at fixed intervals

Right now we have `cloud_recordings` table which stores the following values

```
:frequency, :integer
:storage_duration, :integer
:schedule, EvercamMedia.Types.JSON
```

and this table is used by cloud_recording app to store snapshots in s3. We have a few other apps that need to perform actions at certain interval like timelapse and snapmail. I think we can extend the existing schedules to these apps as well.

I propose to change this table name as `schedules` with the following columns

```
:app_id, integer
:rules, json
:config, json
```

_:rules_
This is just a rename of the existing schedule column with the same JSON type. Better name is welcome. Since we rename the table, I don't like the table and a column in it having the same name.

_:app_id_
relationship to identify which application owns the schedule.

_:action_
a json object storing the configuration for the intended action for the given schedule. For eg.

for cloud recording:

```
{
  app: 'cloud_recording',
  actions: ['s3_upload'],
  frequency: :integer,
  storage_duration: :integer
}
```

for timelapse:

```
{
  app: 'timelapse',
  actions: ['s3_upload', 'generate_movie']
}
```

for snapmail:

```
{
  app: 'snapmail',
  actions: ['send_mail']
  mail_to: 'client@example.com'
}
```

_Note:_ I am not proposing the exact json structure as above. I am sure there is scope for improvement.

`:actions` key in the json structure is a list of actions to be done for every successful snapshot in the given schedule. The other keys store configurations needed for the actions.

**A bit of theory of how the cloud_recordings works now.**

Each camera is having its own GenServer, termed as camera worker. Each of these workers have a Poller which could potentially have many timers. Currently this poller is having a single timer created using cloud_recording schedule. This timer calls the worker for getting a snapshot. When a snapshot is successfully received, we send a GenEvent message 'snapshot_received'. There are many event handlers like s3, db updater etc that can hook into the generated message and perform the necessary action.

**How would this work now with my new proposal?**

We leverage the possibility of Worker's Poller having multiple timers. When timers get a snapshot, they also pass in the configuration from the schedule which contains the list of actions to be made with other details for those actions. So instead of sending GenEvent message like 'snapshot_received', we would now send messages like 'send_mail' or 's3_upload' which are handled by different handlers approriately.

**A note about timers:**
timers are created using erlang's bif :send_after or standard lib :timer module. Both of them are very effective in managing interval bound actions in erlang. For instance, for a snapmail configured at 9am and 6pm, the timer would only wake up only twice per day with no polling/checks made at any other time.

**Comment by shankardevy** [comment]
@mosic @azharmalik3 I would to hear your opinions on the above proposal. 

//cc @marcoherbst

**Comment by marcoherbst** [comment]
Thanks.

To me it's not obvious that we should be building all of this logic into
the scheduler. I've a strong suspicion that when either Snapmail or
Timelapse (but especially Snapmail) is considered in its entirety along
with roadmap to add custom text, logos, and more, it will look more and
more like it should be architected independently of the primary recording
functionality (more like it currently is).

@milos @azharmalik3 , I'd like you to be aware of this when giving your
thoughts.

M



On 8 October 2015 at 13:04, Shankar Dhanasekaran <notifications@github.com>
wrote:

> @mosic <https://github.com/mosic> @azharmalik3
> <https://github.com/azharmalik3> I would to hear your opinions on the
> above proposal.
>
> //cc @marcoherbst <https://github.com/marcoherbst>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/evercam/evercam-media/issues/86#issuecomment-146517659>
> .
>



-- 
Marco Herbst
*Evercam.io <http://www.evercam.io/>* - Cameras Unleashed
Mobile: +353 86 196 0451
Skype: marco.a.herbst

**Comment by mosic** [comment]
Nicely written Shankar. I'm inclined to agree with Marco on this, requirements for other apps aren't specified yet so forcing them to use the same scheduler logic upfront may prove to be more trouble than it is worth. I'd revisit this idea in the future though.

**Comment by marcoherbst** [comment]
It would be great, early next week, to make a decision on timelapse.

I feel like we're possibly approaching the specification of the task from
an "I've just built a scheduling component" perspective as opposed to an "I
need to get some jpegs into a video" perspective. I've made similar
mistakes to this before, hence the caution.

A key issue is to decide how we want to do the on-the-fly timelapse player
( embedded replay of the timelapse to date, incl embedded logo, speed
controls etc. ). I think that while a javascript jpeg player could work,
actually, anything other than a HLS stream will be a compromise which means
creating mpeg ts chuncks on the fly. I'm not sure, but this is my current
thinking.

M

On 8 October 2015 at 16:58, Miloš Mošić <notifications@github.com> wrote:

> Nicely written Shankar. I'm inclined to agree with Marco on this,
> requirements for other apps aren't specified yet so forcing them to use the
> same scheduler logic upfront may prove to be more trouble than it is worth.
> I'd revisit this idea in the future though.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/evercam/evercam-media/issues/86#issuecomment-146591721>
> .
>



-- 
Marco Herbst
*Evercam.io <http://www.evercam.io/>* - Cameras Unleashed
Mobile: +353 86 196 0451
Skype: marco.a.herbst
